utilities/generate_messages_catalog.py

```diff
@@ -0,0 +1,326 @@
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+  <xs:element name="messages">
+    <xs:complexType>
+      <xs:sequence>
+        <xs:element maxOccurs="unbounded" name="message">
+          <xs:complexType>
+            <xs:simpleContent>
+              <xs:extension base="xs:string">
+                <xs:attribute name="code" use="required" type="xs:normalizedString"/>
+                <xs:attribute name="level" use="required" type="xs:token"/>
+                <xs:attribute name="module" type="xs:normalizedString"/>
+                <xs:attribute name="line" type="xs:integer"/>
+                <xs:attribute name="args" type="xs:NMTOKENS"/>
+              </xs:extension>
+            </xs:simpleContent>
+          </xs:complexType>
+        </xs:element>
+      </xs:sequence>
+      <xs:attribute name="variablePrefix" type="xs:string"/>
+      <xs:attribute name="variableSuffix" type="xs:string"/>
+      <xs:attribute name="variablePrefixEscape" type="xs:string"/>
+    </xs:complexType>
+  </xs:element>
+</xs:schema>
+"""
+
+ARELLE_MESSAGES_XML = """<?xml version="1.0" encoding="utf-8"?>
+<messages
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:noNamespaceSchemaLocation="messagesCatalog.xsd"
+    variablePrefix="%("
+    variableSuffix=")s"
+    variablePrefixEscape="" >
+<!--
+This file contains Arelle messages text.   Each message has a code
+that corresponds to the message code in the log file, level (severity),
+args (available through log file), and message replacement text.
+
+(Messages with dynamically composed error codes or text content
+(such as ValidateXbrlDTS.py line 158 or lxml parser messages)
+are reported as "(dynamic)".)
+
+-->
+
+"""
+
+DOC_DIRECTORY = os.sep.join([
+    os.path.split(os.path.dirname(__file__))[0],
+    "arelle", "doc"
+])
+
+
+def _log_function(item):
+    """
+    Handler function for log message types.
+
+    :param item: ast object being inspected for certain properties
+    :type item: :class:`~ast.AST`
+    :return: Returns the descriptor and arg offset of the item.
+    :rtype: tuple (str, int)
+    """
+    level_arg = item.args[0]
+    if isinstance(level_arg, ast.Str):
+        level = level_arg.s.lower()
+    else:
+        any_call_or_name_elements = any(
+            isinstance(element, (ast.Call, ast.Name))
+            for element in ast.walk(level_arg)
+        )
+        if any_call_or_name_elements:
+            level = "(dynamic)"
+        else:
+            level = ', '.join(
+                element.s.lower()
+                for element in ast.walk(level_arg)
+                if isinstance(element, ast.Str)
+            )
+    integer_arg_offset = 1
+    return level, integer_arg_offset
+
+
+FUNC_HANDLER = {
+    "info": ("info", 0),
+    "warning": ("warning", 0),
+    "error": ("error", 0),
+    "exception": ("exception", 0),
+    "log": _log_function
+}
+
+
+def entityEncode(arg):
+    """ Be sure it's a string, vs int, etc, and encode &, <, ". """
```

Comment:
I keep thinking I can replace this with something like the lxml html escaper...I want to try to poof